fix: CTRL/CMD & arrow point drag unbinds both sides (#6459)

mtolmacs opened the pull request:
fix https://github.com/excalidraw/excalidraw/issues/6459

This PR fixes not just #6459 but also addresses the related issue of CTRL/CMD & dragging arrow drag points disconnect ALL arrows in the scene.

## How to test the effect of the PR

### Setup
Create two diamonds and two boxes, then connect one diamond to the other diamond via an arrow, followed by connecting one box to the other via another arrow.

### Before
1) Press the CTRL/CMD key and drag the end marker of the arrow connecting the boxes somewhere ✅ 
2) The box previously connected to the end of the arrow is now disconnected ✅ 
3) When the box connected with the start of the arrow is drag moved, it can be seen that the arrow did not remain connected ❌ 
4) When any of the diamonds are drag moved, the arrow connected to them did not remain connected to either of them ❌ 

### After
1) Press the CTRL/CMD key and drag the end marker of the arrow connecting the boxes somewhere ✅ 
2) The box previously connected to the end of the arrow is now disconnected ✅ 
3) When the box connected with the start of the arrow is drag moved, it remains connected to the arrow ✅ 
4) When the box connected with the end of the arrow is drag moved, it remains connected to the arrow ✅ 
5) When any of the diamonds are moved, the arrow remains connected to both of them ✅ 

## Refactor
In this PR I also propose a function rename in `element/binding.ts`, specifically the `unbindLinearElements` to `filterBindingElementsAndPossiblyUnbind` in order to better reflect what the actual function does. Please let me know if this is inappropriate or need to be proposed in another context.

dwelle commented:
Hey, sorry for the delay.

Works great, except there's one regression we don't have a test for, which is when an arrow is bound (from either side, or both sides) and you start dragging it and then hold cmd/ctrl, it should unbound both ends on release. In this PR it keeps the bindings.

I'd use the opportunity and rewrite the related binding helpers as it's unclear what the difference between `bindOrUnbindSelectedElements` and `filterBindingElementsAndPossiblyUnbind` is (when to use what and what are the nuances). I'd likely aim at making them (more) stateless.

A few notes:

The `filterBindingElementsAndPossiblyUnbind` relies on the (un)binding to have been done upstream before it's called, largely obviating its purpose.

We're also relying on `state.selectedLinearElement` being set to decide whether to bind/unbind in certain interactions, which makes the behavior differ based on whether we already have a selected arrow and move it, or we move it right during selection (i.e. we don't release the initial `pointerdown`), which can result in a different behavior in each case (can't repro rn but I dimly recall there was one bug connected to it; will have to double check).

mtolmacs commented:
Hi @dwelle,

I've fixed the regression where CTRL/CMD + arrowshaft dragging still keeps the arrow connected.

Refactoring seems more complex around the binding logic. Merging `bindOrUnbindSelectedElements ` and `unbindLinearElements` (`filterBindingElementsAndPossiblyUnbind` in my previous commits ) would be possible, but I believe it would make the code significantly less readable. `bindOrUnbindSelectedElements` searches for binding candidates, while `unbindLinearElements` breaks bindings, so the merged function would have 3 level deep branching, which would be harder to follow.

Given that there are more complex refactoring opportunities around binding than just these two helpers, I reverted the helper rename and I would like to focus on fixing the remaining regressions exclusively in this PR, leaving the larger refactor for another PR.

Regarding your notes, hopefully the latter case is not affected as the regression fix is entirely in the scope of the single helper `unbindLinearElements`, which is not called from anywhere else (except `actionFlip.ts` but that doesn't seem to be affected).

As for whether `unbindLinearElements` is superfluous, now it's actually does the heavy lifting of fixing the regression, so until a bigger refactor comes around I'd keep it.

Let me know if this answers all questions and addresses all issues related to #6459

dwelle commented:
Yeah, as discussed async we'll spec up the redesign and do the larger refactor/rewrite then.

As for the issue at hand, that is fixed, and I didn't notice any regressions. So we're good to go, thanks! 👍
